### ISSUE TITLE
Constraints on submissions

### ISSUE THREAD

**Issue by skearnes** [Open-Reaction-Database/ord-schema#135]
As I'm working on the tests for the submission workflow, there are some rules that are obvious to enforce for submissions as well as some that we should discuss. Here's my current list:

1. Submissions must either (a) add new dataset or (b) modify an existing dataset, not both.
1. Submissions must add or modify only a single dataset. Note that to allow for large `Data` messages, submissions may be "sharded" across multiple `Dataset` messages that will be combined into a single dataset during the submission process.
1. New datasets must:
    1. __?__ Not include any dataset or record IDs.
1. Modifications to existing datasets must:
    1. Not change existing dataset or record IDs.
    1. __?__ Not add or remove records (this requires a new dataset).

**Comment by connorcoley** [comment]
I like the idea of these restrictions. Thoughts on your questions:

> 3i: New datasets must not include any dataset or record IDs

We could either require the fields to be blank or simply overwrite them; I'm okay with the latter.

> 4ii: Modifications to existing datasets must not add or remove records

There may be cases where either (a) a submission mistakenly excluded a reaction that should be part of a dataset, perhaps because they were extracting data from a published paper, or (b) a chemist decides to run an additional plate or two from an HTE experiment that they already uploaded.

**Comment by skearnes** [comment]
Update on this:

It's quite difficult to check whether a dataset or reaction ID already existed or was added by the automated pipeline, especially if there is more than one commit from the submitter in a PR.

The biggest risk IMO is that we have duplicated IDs. To protect against this, we could include a check that verifies there are no duplicated IDs across the entire database. I'd suggest we start with a brute force approach and do fancier index-type things as needed for scaling.

I'm not entirely sure how to automatically identify cases where IDs are _changed_; looking at raw diffs is fraught with peril. If we enforce mutual exclusivity between adding vs. editing we can check that the set of IDs before and after the change is the same, which isn't perfect (i.e. doesn't catch shuffling IDs) but might be enough.
